### PR TITLE
hide ranks without search match #774

### PR DESCRIPTION
**Pull request by GiacomoCaruso**
This will hide the rank listings when no users are found to match the current search string. See issue #774 
My goal was to reduce modifications to DOM elements by hiding rank listings instead of hiding single usernames.

### CODE REVIEW

**Comment by MaxLeiter** [comment]
Awesome! Just a warning, the builds are failing due to the styling errors here:
https://travis-ci.org/thelounge/lounge/jobs/195170214

**Comment by GiacomoCaruso** [comment]
Thank you for the warning, Max! I've just fixed the styling issues.

**Comment by MaxLeiter** [comment]
If you could also squash your commits that'd be awesome - just run `npm rebase -i HEAD~2` and change the second line to say `f` instead of `pick` at the beginning. Then save and `git push -f`. This will turn your two commits into one. Since 2.2.0 is right around the corner, I don't think this will be merged/reviewed for a few days, but thanks for the work :)

EDIT: may want to see my comments and change if you want, then commit + squash (change HEAD\~2 -> HEAD~3 to squish 3 commits)

File: client/js/lounge.js [hunk 1]
```diff
@@ -1105,13 +1105,35 @@ $(function() {
...
-				btn.show();
+		// Find all the categories of users and loop through them.
+		names.find(".user-mode").each(function() {
+			// Declare array to store names to hide, in case the whole category can't be hidden.
+			var namesToHide = [];
+			var userModeDiv = $(this);
+			// Declare boolean variable to establish if the current category should be hidden.
+			var isDivToHide = true;
+			// Find and loop through each name.
+			userModeDiv.find(".user").each(function() {
```
Comment (MaxLeiter) [inline client/js/lounge.js]: arrow notation?

File: client/js/lounge.js [hunk 2]
```diff
@@ -1105,13 +1105,35 @@ $(function() {
...
+				}
+			});
+			// If possible, hide the whole category, otherwise hide the "names to hide".
+			if (isDivToHide) {
+				userModeDiv[0].style.display = "none";
 			} else {
-				btn.hide();
+				userModeDiv[0].style.display = "block";
+				var i;
+				for (i in namesToHide) {
```
Comment (MaxLeiter) [inline client/js/lounge.js]: I'd personally prefer a for-loop but not really necessary imo

File: client/js/lounge.js [hunk 3]
```diff
@@ -1105,13 +1105,35 @@ $(function() {
 	chat.on("input", ".search", function() {
 		var value = $(this).val().toLowerCase();
 		var names = $(this).closest(".users").find(".names");
-		names.find(".user").each(function() {
-			var btn = $(this);
-			var name = btn.text().toLowerCase().replace(/[+%@~]/, "");
-			if (name.indexOf(value) > -1) {
-				btn.show();
+		// Find all the categories of users and loop through them.
+		names.find(".user-mode").each(function() {
```
Comment (MaxLeiter) [inline client/js/lounge.js]: arrow notation?

Comment (MaxLeiter) [inline client/js/lounge.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions

Comment (GiacomoCaruso) [inline client/js/lounge.js]: I have no idea why, but arrow functions don't seem to work in this case. I can't even type in the search box, but no errors are thrown anywhere. Not in the tests, not in the console. Maybe I am doing something wrong? I simply replaced `function()` with `() =>` and kept the rest the same.

File: client/js/lounge.js (hunk 2 above)
Comment (GiacomoCaruso) [inline client/js/lounge.js]: Right, I just found out that the for/in loop is slower and it is meant for objects, not arrays. My bad.

**Comment by GiacomoCaruso** [comment]
First time using git rebase and I did something wrong. Somehow it looks like I re-pushed a previous commit of another contributor. I am trying to figure out how to fix it, but I am not sure I can.

**Comment by xPaw** [comment]
I think this PR is rendered unnecessary by #856.

**Comment by PugaBear** [comment]
The gif in #856 still shows the "Operators" list header in the search results even though there were no matches with the Operator rank. This is what this commit was designed to fix. However this might need to be changed to be compatible with that commit. @xPaw

**Comment by GiacomoCaruso** [comment]
Yeah, that's true. If it's needed, I could give it another try after you re-design the search feature with fuzzy.

**Comment by xPaw** [comment]
@PugaBear that gif does not include changes which I commited. My change renders search results into a separate container.

**Comment by xPaw** [comment]
Closing this in favour of #856 since that PR re-engineers the same piece of code and is more likely to be merged.

**Pull request closed without merging**